fedimint-core/src/encoding/mod.rs

```diff
@@ -611,6 +614,139 @@ impl Decodable for Cow<'static, str> {
+        raw: Vec<u8>,
+    },
+    Decoded(T),
+}
+
+impl<T> DynRawFallback<T>
+where
+    T: Decodable + 'static,
+{
+    pub fn expect_decoded(self) -> T {
+        match self {
+            DynRawFallback::Raw { .. } => panic!("Expected decoded value"),
+            DynRawFallback::Decoded(v) => v,
+        }
+    }
+
+    pub fn expect_decoded_ref(&self) -> &T {
+        match self {
+            DynRawFallback::Raw { .. } => panic!("Expected decoded value"),
+            DynRawFallback::Decoded(v) => v,
+        }
+    }
+
+    /// Attempt to re-decode raw values with new set of of `modules`
+    ///
+    /// In certain contexts it might be necessary to try again with
+    /// a new set of modules.
+    pub fn redecode_raw(
+        self,
+        decoders: &ModuleDecoderRegistry,
+    ) -> Result<Self, crate::encoding::DecodeError> {
+        Ok(match self {
+            DynRawFallback::Raw {
+                module_instance_id,
+                raw,
+            } => match decoders.get(module_instance_id) {
+                Some(decoder) => DynRawFallback::Decoded(decoder.decode(
+                    &mut &raw[..],
+                    module_instance_id,
+                    decoders,
+                )?),
+                None => DynRawFallback::Raw {
+                    module_instance_id,
+                    raw,
+                },
+            },
+            DynRawFallback::Decoded(v) => DynRawFallback::Decoded(v),
+        })
+    }
+}
+
+impl<T> From<T> for DynRawFallback<T> {
+    fn from(value: T) -> Self {
+        Self::Decoded(value)
+    }
+}
+
+impl<T> Decodable for DynRawFallback<T>
+where
+    T: Decodable + 'static,
+{
+    fn consensus_decode<R: std::io::Read>(
+        reader: &mut R,
+        decoders: &ModuleDecoderRegistry,
+    ) -> Result<Self, crate::encoding::DecodeError> {
+        let module_instance_id =
+            fedimint_core::core::ModuleInstanceId::consensus_decode(reader, decoders)?;
+        Ok(match decoders.get(module_instance_id) {
+            Some(decoder) => {
+                DynRawFallback::Decoded(decoder.decode(reader, module_instance_id, decoders)?)
+            }
+            None => {
+                // since the decoder is not available, just read the raw data
+                let total_len = u64::consensus_decode(reader, decoders)? as usize;
+                let mut buf = Vec::with_capacity(512);
+                while buf.len() < total_len {
+                    let prev_len = buf.len();
+                    let new_len = std::cmp::min(total_len, prev_len.saturating_add(512));
+                    buf.resize(new_len, 0u8);
+                    reader
+                        .read_exact(&mut buf[prev_len..])
+                        .map_err(|e| fedimint_core::encoding::DecodeError::new_custom(e.into()))?;
+                }
+                Self::Raw {
+                    module_instance_id,
+                    raw: buf,
+                }
```

Comment:
#2742 introduced quite similar code iirc, could it be deduplicated?

Reply:
I think they are already, unless I'm missing something.

Reply:
Ah, maybe I missed that reviewing commit by commit